### PR TITLE
ARROW-12263: [Dev][Packaging] Move Crossbow to Archery

### PR DESCRIPTION
**Pull request by kszucs**
Main highlights:
- prepared for unittesting
- introduced jinja macros to reduce code duplication for crossbow related snippets, like arrow checkout or release uploading
- the tasks.yml is jinja templated now
- fix error the error when a task considered successful despite no build is being scheduled

### CODE REVIEW

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
@github-actions crossbow submit -g wheel

**Comment by kszucs** [comment]
Submitted nightly jobs manually: https://github.com/ursacomputing/crossbow/branches/all?query=build-141

**Comment by kszucs** [comment]
Submitted nightly jobs manually: https://github.com/ursacomputing/crossbow/branches/all?query=build-187

**Review by kou — APPROVED** [review]
+1

I can't review all codes but it seems to be working fine.

**Comment by kszucs** [comment]
I need to adjust the comment bot which I'm testing on my fork.

**Review by kszucs — COMMENTED** [review]
+1

File: dev/archery/archery/bot.py [hunk 1]
```diff
@@ -248,81 +167,94 @@ def actions(ctx):
...
+            crossbow_repo=crossbow_repo,
+            pull_request=pull_request,
         )
+        # load available tasks configuration and groups from yaml
+        config = Config.load_yaml(arrow.path / "dev" / "tasks" / "tasks.yml")
+        config.validate()
+
+        # initialize the crossbow build's target repository
+        target = Target.from_repo(arrow, version=arrow_version,
+                                  remote=pull_request.base.repo.clone_url)
```
Comment (kou) [inline dev/archery/archery/bot.py:243]: It seems that `pull_request.base` always refers `apache/arrow` not `${USER}/arrow`: https://github.com/apache/arrow/pull/9977#issuecomment-817082661
Should w use `pull_request.head` here?

Comment (kou) [reply dev/archery/archery/bot.py:243]: @kszucs ↑

Comment (kszucs) [reply dev/archery/archery/bot.py:243]: We should, sorry about that! The fix is at https://github.com/apache/arrow/pull/9978

Comment (kou) [reply dev/archery/archery/bot.py:243]: Thanks!
It works now!

**Pull request closed without merging**